### ISSUE TITLE
Variables with initial underscores

### ISSUE THREAD

**Issue by jjatria** [praat/praat#17]
When writing procedures, I often find it useful to differentiate between procedures for internal use, and procedures for general use. Taking a hint from Perl and others, I like to preppend internal procedures with an initial underscore (`_`).

Variables in Praat are not allowed to begin with an underscore. But since "local" variables in procedures take their name from the name of the procedure they belong to, and procedures _can_ be named with an initial underscore, this results in all sorts of fun with Praat variables:

~~~~
procedure _test ()
  .local = 1
endproc
@_test()

# Works
a = '_test.local'
appendInfoLine: a

# Doesn't
asserterror Unknown symbol:'newline$'« _
a = _test.local
~~~~

Is this desired behaviour? I've learned now to navigate around the problem as illustrated above, but it feels a bit inconsistent. This is one of the reasons I sometimes still use variable substitution.

**Comment by PaulBoersma** [comment]
Yes, this is partly desired behaviour. With the single quotes, you can have variables with any content, for instance those weird ones that could come from forms, such as '%_correct' and the like.
I understand why you could want to have a procedure name that starts with an underscore, namely in included files: for procedures that are used only within the included file and are thus not exported to the including script. Still, that's namespace pollution, so prepending the name of the included "module" seems safer (or underscore plus module).

**Comment by jjatria** [comment]
Thanks! That's very good to know. And since there is a workaround I don't even think the behaviour should necessarily change: it makes sure that people who choose to do this know what they are doing.

I've been doing something like what you say about prepending names in the case of [`vieweach`](https://gitlab.com/cpran/plugin_vieweach), which "exports" `for_each` and `view_each`. These are the ones with "predefinable" procedures, which are named like that to make that connection explicit: `view_each.finally`, `for_each.at_end_iteration`, etc.

So far, I've only had to use one "internal" procedure ([`_try_editor`](https://gitlab.com/cpran/plugin_vieweach/blob/cd6dc2267730e7c736cb31214d501209b9f9ac49/procedures/view_each.proc#L162)). That could be renamed like you say to something even _less likely_ like `_vieweach.try_editor`, but that's of course a completely different conversation.

Maybe this should have been an email to the list instead? :/

**Comment by PaulBoersma** [comment]
This seems to be about coding style. Let's have a page in the manual about that. The suggestion is also to have importable modules rather than includable source files. And perhaps we'd like to have _ to start names of variables that are truly local (on a stack, so that procedures can be called recursively).

**Comment by jjatria** [comment]
That sounds great. Particularly the "importable modules" as opposed to included files. I'm very interested in how that would be implemented, because of what we've already talked about regarding the limitations of includes with fixed paths.

In what you are thinking, how would the two ("modules" v. "plugins"?) differ from one another?

Truly local variables (and easy recursive procedures) would be [really great](https://gist.github.com/jjatria/cb03a2afe1691145dc01). On that, how would they be scoped? Removed from the stack at the corresponding `endproc`?

But I think having truly local variables will make it even more necessary to implement return values for procedures, because until now it is precisely because variables are not local that we can get values from procedures. How would truly local variables fit into this model (other than by ending scripts with a `.return = _return` line)?

**Comment by jjatria** [comment]
A side effect of this I recently noticed. With an example like

``` praat
procedure _test ()
  .secret$    = "exposed"
  .secret$[1] = "secret!"
endproc
```

The value of the `_test.secret$` variable can be interpolated with `'_test.secret$'`, but the value of `_test.secret$[1]` is (as of Praat 6.0.29 at least) **entirely** inaccessible.

* It cannot be accessed directly because of the leading underscore
* It cannot be interpolated because of the index

Food for thought. :thinking:
